dev_scripts/qa.py

```diff
@@ -0,0 +1,448 @@
+            if " " in arg:
+                cmd_str += " '" + arg + "'"
+            else:
+                cmd_str += " " + arg
+
+        logger.debug("Running %s", cmd_str)
+        try:
+            subprocess.run(args, check=True)
+            logger.debug("Successfully ran %s", cmd_str)
+        except subprocess.SubprocessError as e:
+            logger.debug("Failed to run %s", cmd_str)
+            raise
+
+    def try_run(self, *args):
+        """Try to run a command and return True/False, depending on the result."""
+        try:
+            self.run(*args)
+            return True
+        except subprocess.SubprocessError as e:
+            return False
+
+    def describe(self, *msgs, ref=None):
+        """Print a task's description.
+
+        Log one message per line, and optionally allow a URL as a reference.
+        """
+        for msg in msgs:
+            logger.info(msg)
+        if ref:
+            logger.info("For reference, see: %s" % ref)
+
+    def prompt(self, *msgs, ref=None, prompt=None):
+        """Print the task's description, and prompt the user for an action."""
+        self.describe(*msgs, ref=ref)
+        prompt = prompt or "Press Enter once you completed this step to continue: "
+        if self.skip_manual:
+            logger.info("Skipping manual tasks, as instructed")
+            return None
+        else:
+            return input(prompt)
+
+    def qa_scenarios(self, skip=None):
+        """Common prompt for the QA scenarios.
+
+        This method suggests to the user a way to check the QA scenarios, in a
+        Dangerzone environment. Then, it iterates through them, prints their
+        description, and asks the user if they pass.
+        """
+        skip = skip or []
+        REF_QA = "https://github.com/freedomofpress/dangerzone/blob/49b7736cb49d8b03d44c66a3903ba4b9ba1f9171/RELEASE.md#scenarios"
+        self.describe(
+            "Beginning QA scenarios",
+            "You can execute into a test environemnt with:",
+            "",
+            f"    cd {self.src}",
+            f"    ./dev_scripts/env.py --distro {self.DISTRO} --version"
+            f" {self.VERSION} run -g bash",
+            "",
+            "and run either `dangerzone` or `dangerzone-cli`",
+            ref=REF_QA,
+        )
+        for num, scenario in enumerate(QA_SCENARIOS):
+            self.describe(scenario)
+            if num + 1 in skip:
+                self.describe("Skipping for the current platform")
+            else:
+                self.prompt("Does it pass?")
+        logger.info("Successfully completed QA scenarios")
+
+    def start(self):
+        """Start the QA tests for a platform."""
+        raise NotImplementedError("No steps have been defined")
+
+
+# TODO: Test this class on Windows thorougly, and finish it.
+class QAWindows(QABase):
+    """Class for the Windows QA tasks."""
+
+    REF_BUILD = (
+        "https://github.com/freedomofpress/dangerzone/blob/main/BUILD.md#windows"
```

Comment:
First of all, I tried looking for a Python project that does something like the above, but I didn't find anything. Either my Google-fu have failed me, or no (popular) project exists that takes a stab at the problem of code deviation from a source of truth.

In any case, your solution is very cool, and we should definitely adopt it. I see some minor issues though, that I'd like to discuss:

Relying on the Git revision is problematic, because you can't reference a Git hash in the same Git commit. You have to do so in a subsequent commit. Even then, it's not easy to do so from a PR, because the hash may change, e.g. due to a rebase, fixups, etc.

I think we need to use SHA-256 instead for these blocks. They are a bit more difficult to compute (e.g., run `sed -n 3,47p BUILD.md | sha256sum` for the Debian/Ubuntu section on `BUILD.md`), but they don't have these drawbacks.

Reply:
After a sync discussion, we'll do the following:

1. Move check for diffs on instance method instead of main on the Reference class.
2. Specify the whole section in the Reference (this will be verbose)
3. Introduce qa.py --check-references
4. Run qa.py --check-references on our lint checks.